db/db.ts

```diff
@@ -310,3 +317,110 @@ export const getHomepageId = (
             AND published = TRUE`
     ).then((result) => result?.id)
 }
+
+export const getImageMetadataByFilenames = async (
+    knex: KnexReadonlyTransaction,
+    filenames: string[]
+): Promise<Record<string, ImageMetadata & { id: number }>> => {
+    if (filenames.length === 0) return {}
+    const rows = await knexRaw<ImageMetadata & { id: number }>(
+        knex,
+        `-- sql
+        SELECT
+            id,
+            googleId,
+            filename,
+            defaultAlt,
+            updatedAt,
+            originalWidth,
+            originalHeight
+        FROM
+            images
+        WHERE filename IN (?)`,
+        [filenames]
+    )
+    return keyBy(rows, "filename")
+}
+
+export const getPublishedGdocPosts = async (
+    knex: KnexReadonlyTransaction
+): Promise<DbEnrichedPostGdoc[]> => {
+    return knexRaw<DbRawPostGdoc>(
+        knex,
+        `-- sql
+        SELECT
+        g.breadcrumbs,
+        g.content,
+        g.createdAt,
+        g.id,
+        g.markdown,
+        g.publicationContext,
+        g.published,
+        g.publishedAt,
+        g.revisionId,
+        g.slug,
+        g.updatedAt
+    FROM
+        posts_gdocs g
+    WHERE
+        g.published = 1
+        AND g.content ->> '$.type' IN (:types)
+        AND g.publishedAt <= NOW()
+    ORDER BY g.publishedAt DESC`,
+        {
+            types: [
+                OwidGdocType.Article,
+                OwidGdocType.LinearTopicPage,
+                OwidGdocType.TopicPage,
+                OwidGdocType.AboutPage,
+            ],
+        }
+    ).then((rows) => rows.map(parsePostsGdocsRow))
+}
+
+export const getPublishedGdocPostsWithTags = async (
+    knex: KnexReadonlyTransaction
+): Promise<DBEnrichedPostGdocWithTags[]> => {
+    return knexRaw<DBRawPostGdocWithTags>(
+        knex,
+        `-- sql
+        SELECT
+        g.breadcrumbs,
+        g.content,
+        g.createdAt,
+        g.id,
+        g.markdown,
+        g.publicationContext,
+        g.published,
+        g.publishedAt,
+        g.revisionId,
+        g.slug,
+        g.updatedAt,
+        JSON_ARRAYAGG(
+          JSON_OBJECT(
+            'id', t.id,
+            'name', t.name,
+            'slug', t.slug
+        )) AS tags
```

Comment:
If you change it to a left join, then we would also want nicer null array values:
```suggestion
        JSON_ARRAYAGG(
          IF(t.id IS NULL, NULL, JSON_OBJECT(
            'id', t.id,
            'name', t.name,
            'slug', t.slug
        ))) AS tags
```

Note that this then may result in something like `tags: [null]`, which you need to filter out afterwards (for example, using `excludeNullish`). But at least it won't result in the very awkward `{ id: null, name: null, slug: null }`

Reply:
You can do it slightly differently and pull the if up one level then you don't have to do excludeNullish:

```suggestion
        if( count(t.id) = 0, json_array(), JSON_ARRAYAGG(
          JSON_OBJECT(
            'id', t.id,
            'name', t.name,
            'slug', t.slug
        ))) AS tags
```